content/md/posts/clojure-which-execution-option-to-use.md

```diff
@@ -0,0 +1,217 @@
+
+The exec-opts command line flags have evolved to enable these features, so lets explore those flags and show how each flag is typically used.
+
+<!-- more -->
+
+## Summary of use
+
+In very simple terms, the Clojure CLI tools flags are used as follows
+
+`-A` - [Deprecated] use `-M` instead
+
+`-M` for running the `-main` function of the current project, using the `-m` flag to specify the namespace containing `-main`
+
+`-X` for running any fully qualified function from the current project
+
+`-P` a dry run, downloading all dependencies
+
+`-T` running a separate tool
+
+> The tools flag, `-T` removes the need for other aliases to run tools and eventually all community tools should move to use the -T approach.  However, some tools still need to be run using the `-M` or `-X` flag (as the `-T` flag is quite new).
+
+
+## clojure.main -M
+
+[`clojure.main` namespace](https://clojure.org/reference/repl_and_main) has been the way Clojure code was run (including a REPL) for most of its history.  This is now evolving with the addition of [clojure.exec](#clojure-exec--X).
+
+`clojure.main/main` function looks for a `-main` function in the given namespace (`-m fully-qualified.namespace`) or if no main namespace is found then a REPL session is run.  This is the approach Leiningen has always used and still uses today.
+
+Running a project using Clojure CLI tools without any additional aliases on the command line
+
+```bash
+clojure -M -m practicalli.sudoku-solver
+```
+
+`-M` flag instructs Clojure CLI tools to use `clojure.main` to run the Clojure code.
+
+`-m` flag is an argument to `clojure.main` which specifies the namespace to search for a `-main` function.
+
+The same approach is taken when including an alias, in this example `:alias/name`.
+
+```bash
+clojure -M:alias/name -m practicalli.sudoku-solver
+```
+
+This command will merge values defined in the alias for `:extra-paths` and `extra-deps` into the call.
+
+> When the command line includes the `-m` flag with a namespace it will over-ride the `:main-opts` value if defined in the alias
+
+
+Adding a `:project/run` alias to the project `deps.edn` file provides a simpler way to run the project on the command line
+
+```clojure
+  :project/run
+  {:main-opts ["-m" "practicalli.sudoku-solver"]}
+
+```
+
+Now the project code can be run using the simple command line form
+
+```bash
+clojure -M:project/run
+```
+
+
+### Chaining aliases together
+
+Alises can be used together by chaining their names on the command line
+
+```bash
+clojure -M:env/dev:env/test:lib/cider:repl/rebel
+```
+
+Using several aliases on the command line will merge their `:extra-paths` and `:extra-deps` values.
+
+The `:main-opts` values are not merged and the value from the last `:main-opts` in the alias chain is used with `clojure.main`
+
+Again, if the command line includes the `-m` flag with a namespace, then that namespace is passed to `clojure.main`, ignoring the values from the aliases.
```

Comment:
More than this: if your command-line includes `-i` or `-e` options those also replace the `:main-opts` in any aliases, since they are all consider main opts together.